pkg/cdi/cache.go

```diff
@@ -277,3 +339,113 @@ func (c *Cache) GetSpecDirectories() []string {
 	copy(dirs, c.specDirs)
 	return dirs
 }
+
+// Our fsnotify helper wrapper.
+type watch struct {
+	w       *fsnotify.Watcher
+	watched map[string]bool
+}
+
+// Start watching a set of directories for relevant changes.
+func (w *watch) start(dirs []string, m *sync.Mutex, refresh func() error) error {
+	var (
+		dir string
+		err error
+	)
+
+	w.watched = make(map[string]bool)
+
+	w.w, err = fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "failed to create Spec dir watch")
+	}
+
+	for _, dir = range dirs {
+		w.watched[dir] = false
+	}
+
+	w.update()
+	go w.watch(m, refresh)
+	refresh()
+
+	return nil
+}
+
+// Stop watching directories.
+func (w *watch) stop() {
+	if w.w == nil {
+		return
+	}
+
+	w.w.Close()
+	w.watched = nil
+}
+
+// Update watch with pending/missing or removed directories.
+func (w *watch) update(removed ...string) bool {
+	var (
+		dir    string
+		done   bool
+		err    error
+		update bool
+	)
+
+	for dir, done = range w.watched {
+		if done {
+			continue
+		}
+		err = w.w.Add(dir)
+		switch {
```

Comment:
Is there a particular reason for a switch statement here instead of an `if-else` or `if-continue`?